### ISSUE TITLE
max_allowed_packet does not take into account packet header

### ISSUE THREAD

**Issue by Diggsey** [blackbeam/rust-mysql-simple#64]
If you send packets of exactly `max_allowed_packet`, the `write_packet` function will accept it, but the server will close the connection because the packet is too large.

This can be tested by using `LOAD DATA LOCAL INFILE` (without a handler) for any sufficiently large file (twice `max_allowed_packet` should do it).

**Comment by blackbeam** [comment]
Looks like mysqld issue to me.

Consider the following situation:
User wants to write local infile data to a *server* with a *driver*. Server reports that `max_allowed_packet`, for instance, is 33 554 432 bytes (32MB), so user composed exactly 33 554 432 bytes file. Local infile data should be sent as one packet to a server, but in this case packet size is greater than maximum possible payload in one communication packet, so driver [splits it into chunks](https://dev.mysql.com/doc/internals/en/sending-more-than-16mbyte.html) of 16 777 215 bytes, then sends 3 packets. First packet with payload of size 16 777 215 bytes, second packet with payload of size 16 777 215 bytes and third packet with payload of size 2 bytes. Resulting payload size becomes exactly 33 554 432 bytes, but server drops the connection and prints `Got a packet bigger than 'max_allowed_packet' bytes` to stderr. Driver can't do anything with this because it just sent what user asked it to send and it can't just drop a part of this data, and user just shocked because `max_allowed_packet` reported by a server turned to be a lie.

And also it is not always lie because server accepts data size == `max_allowed_packet` if `max_allowed_packet` == 16MB (one communication packet with payload of 16777215 and one communication packet with payload of 1 byte).

**Comment by Diggsey** [comment]
@blackbeam In this case it is the `Write` implementation of `LocalInfile`, not the user, which is splitting it up into chunks of `max_allowed_packet` - in my tests, using `max_allowed_packet/2` instead completely fixes the problem, allowing arbitrarily large files to be sent.

However, we should be able to calculate the optimal size of packet to send (ie. `max_allowed_packet - <some_small_number>`). I'm just not confident enough in my understanding of the protocol to say what that small number should be.

**Comment by blackbeam** [comment]
> In this case it is the Write implementation of LocalInfile, not the user, which is splitting it up into chunks of max_allowed_packet - in my tests, using max_allowed_packet/2 instead completely fixes the problem, allowing arbitrarily large files to be sent.

I see that current implementation of `Write` trait for `LocalInfile` could break [Protocol::LOCAL_INFILE_Data](https://dev.mysql.com/doc/internals/en/com-query-response.html#packet-Protocol::LOCAL_INFILE_Data) because zero length packets may arise before final zero length packet in cases if data length % MAX_PAYLOAD_LEN == 0.

I also found out that myslq will reject following three packets with total size of payload == 32MB (`LocalInfile` buffer length is `MAX_PAYLOAD_LEN`):
```
Chunk len: 16777215 << first packet
Chunk len: 16777215
Chunk len: 2 << last packet
Chunk len: 0 << zero length packet terminating local infile data
```

but will accept following three packets with total size of payload == 32MB (`LocalInfile` buffer length is `MAX_PAYLOAD_LEN - 1`):
```
Chunk len: 16777214
Chunk len: 16777214
Chunk len: 4
Chunk len: 0
```

so i will set `LocalInfile` buffer to `MAX_PAYLOAD_LEN - 1`.

It also looks like it does not take `max_allowed_packet` to account when handling local infile data but only if chunk payload is smaller than  MAX_PAYLOAD_LEN.

**Comment by blackbeam** [comment]
Published as v7.1.2
